File: backend/src/cti_app/api/publication.py
"""HTTP API for the append-only edition publication review checkpoint."""

from __future__ import annotations

from typing import Annotated, NoReturn
from uuid import UUID

from fastapi import APIRouter, HTTPException, Request, Response, status
from pydantic import BaseModel, ConfigDict, Field, model_validator

from cti_app.api.production import ProductionReconciliationView, reconciliation_view
from cti_app.application.edition_publication import (
    EditionPublicationService,
    EditionReleaseStatus,
    PublicationAcceptanceError,
    PublicationManifestNotFoundError,
)
from cti_app.application.edition_release_materialization import (
    EditionReleaseMaterializationError,
)
from cti_app.application.edition_review import (
    EditionReview,
    EditionReviewItemNotFoundError,
    EditionReviewNotFoundError,
    EditionReviewService,
    EditionReviewStatusError,
    InvalidReviewDocumentError,
    InvalidReviewReasonError,
    ReviewItemStaleError,
)
from cti_app.application.identity import IdentityProvider
from cti_app.domain.jobs import JobStatus
from cti_app.domain.production import SubjectProductionStage, SubjectProductionStatus
from cti_app.domain.publication_review import PublicationDecision, PublicationReviewDecision
from cti_app.logging import get_correlation_id

router = APIRouter(prefix="/api", tags=["publication"])
_SHA256_PATTERN = r"^[0-9a-f]{64}$"


class ReviewRunRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    production_run_id: UUID
    pipeline_generation: Annotated[int, Field(ge=0)]


class ReviewDocumentRequest(ReviewRunRequest):
    document_artifact_id: UUID
    document_artifact_version: Annotated[int, Field(ge=1)]
    document_input_hash: Annotated[str, Field(pattern=_SHA256_PATTERN)]


class IncludeReviewRequest(ReviewDocumentRequest):
    reason: str | None = Field(default=None, max_length=500)


class ExcludeReviewRequest(ReviewRunRequest):
    document_artifact_id: UUID | None = None
    document_artifact_version: Annotated[int | None, Field(ge=1)] = None
    document_input_hash: Annotated[str | None, Field(pattern=_SHA256_PATTERN)] = None
    reason: Annotated[str, Field(min_length=1, max_length=500)]

    @model_validator(mode="after")
    def document_identity_is_atomic(self) -> ExcludeReviewRequest:
        values = (
            self.document_artifact_id,
            self.document_artifact_version,
            self.document_input_hash,
        )
        if any(value is None for value in values) and any(value is not None for value in values):
            raise ValueError("document identity must be complete or empty")
        return self


class ReviewItemView(BaseModel):
    position: int
    subject_id: UUID
    title: str
    run_id: UUID
    pipeline_generation: int
    run_status: SubjectProductionStatus
    document_artifact_id: UUID | None
    document_artifact_version: int | None
    document_input_hash: str | None
    error_code: str | None
    error_message: str | None
    effective_decision: PublicationDecision | None
    effective_decision_id: UUID | None
    included: bool
    blocking: bool
    rejected_indicator_count: int = 0
    rejected_rule_count: int = 0
    published_rule_count: int = 0
    # The frontend must never infer the retry policy from an error message:
    # ``can_retry`` and ``requires_reconciliation`` are mutually exclusive and
    # each names exactly one operator action.
    can_retry: bool
    retry_stage: SubjectProductionStage | None
    requires_reconciliation: bool = False
    reconciliation: ProductionReconciliationView | None = None


class EditionReviewView(BaseModel):
    edition_id: UUID
    items: list[ReviewItemView]
    can_accept: bool


class ReviewDecisionView(BaseModel):
    id: UUID
    edition_id: UUID
    subject_id: UUID
    production_run_id: UUID
    pipeline_generation: int
    document_artifact_id: UUID | None
    document_artifact_version: int | None
    document_input_hash: str | None
    decision: PublicationDecision
    actor_id: str
    reason: str | None
    occurred_at: str


class PublicationAcceptView(BaseModel):
    edition_id: UUID
    edition_status: str
    manifest_id: UUID
    manifest_sha256: str
    edition_version: int
    batch_id: UUID
    job_id: UUID | None
    job_dispatched: bool


class EditionReleaseView(BaseModel):
    edition_id: UUID
    edition_status: str
    manifest_id: UUID | None
    manifest_sha256: str | None
    release_id: UUID | None
    json_available: bool
    markdown_available: bool
    docx_available: bool
    published_at: str | None
    assembly_job_id: UUID | None
    assembly_status: JobStatus | None
    assembly_error_code: str | None
    assembly_error_message: str | None
    can_retry_assembly: bool


class EditionReleaseMaterializationView(BaseModel):
    edition_id: UUID
    materialized: bool


def _service(request: Request) -> EditionReviewService:
    configured = getattr(request.app.state, "edition_review_service", None)
    return configured or EditionReviewService(request.app.state.uow_factory)


def _publication_service(request: Request) -> EditionPublicationService:
    configured = getattr(request.app.state, "edition_publication_service", None)
    if configured is None:
        configured = EditionPublicationService(
            request.app.state.uow_factory,
            request.app.state.production_artifact_store,
            job_service=getattr(request.app.state, "job_service", None),
            job_dispatcher=getattr(request.app.state, "job_dispatcher", None),
        )
    return configured


async def _actor_id(request: Request) -> str:
    provider: IdentityProvider = request.app.state.identity_provider
    return (await provider.current()).actor_id


@router.get("/editions/{edition_id}/review", response_model=EditionReviewView)
async def get_edition_review(edition_id: UUID, request: Request) -> EditionReviewView:
    try:
        return _review_view(await _service(request).get(edition_id))
    except Exception as exc:
        _raise_review_error(exc)


@router.post(
    "/editions/{edition_id}/review/items/{subject_id}/include",
    response_model=ReviewDecisionView,
)
async def include_review_item(
    edition_id: UUID,
    subject_id: UUID,
    payload: IncludeReviewRequest,
    request: Request,
) -> ReviewDecisionView:
    return await _decide(
        edition_id,
        subject_id,
        payload,
        request,
        PublicationDecision.INCLUDE,
    )


@router.post(
    "/editions/{edition_id}/review/items/{subject_id}/exclude",
    response_model=ReviewDecisionView,
)
async def exclude_review_item(
    edition_id: UUID,
    subject_id: UUID,
    payload: ExcludeReviewRequest,
    request: Request,
) -> ReviewDecisionView:
    return await _decide(
        edition_id,
        subject_id,
        payload,
        request,
        PublicationDecision.EXCLUDE,
    )


@router.post(
    "/editions/{edition_id}/publication/accept",
    response_model=PublicationAcceptView,
    status_code=status.HTTP_202_ACCEPTED,
)
async def accept_edition_publication(edition_id: UUID, request: Request) -> PublicationAcceptView:
    try:
        result = await _publication_service(request).accept(
            edition_id,
            actor_id=await _actor_id(request),
            correlation_id=get_correlation_id(),
        )
        return PublicationAcceptView(
            edition_id=result.manifest.edition_id,
            edition_status=result.edition_status.value,
            manifest_id=result.manifest.id,
            manifest_sha256=result.manifest.content_sha256,
            edition_version=result.manifest.edition_version,
            batch_id=result.manifest.batch_id,
            job_id=result.job_id,
            job_dispatched=result.job_dispatched,
        )
    except Exception as exc:
        _raise_publication_error(exc)


@router.get("/editions/{edition_id}/release", response_model=EditionReleaseView)
async def get_edition_release(edition_id: UUID, request: Request) -> EditionReleaseView:
    try:
        release = await _publication_service(request).release_status(edition_id)
        return _release_view(release)
    except Exception as exc:
        _raise_publication_error(exc)


@router.post(
    "/editions/{edition_id}/release/materialize",
    response_model=EditionReleaseMaterializationView,
)
async def materialize_edition_release(
    edition_id: UUID, request: Request
) -> EditionReleaseMaterializationView:
    """Rebuild only the disposable release projection from canonical storage."""
    materializer = getattr(request.app.state, "edition_release_rematerializer", None)
    if materializer is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={"code": "release_materialization_unavailable"},
        )
    await _actor_id(request)
    try:
        await materializer.materialize(edition_id)
    except EditionReleaseMaterializationError as exc:
        code = str(exc)
        status_code = (
            status.HTTP_404_NOT_FOUND
            if code in {"edition_not_found", "manifest_not_found", "edition_release_not_found"}
            else status.HTTP_409_CONFLICT
        )
        raise HTTPException(status_code=status_code, detail={"code": code}) from exc
    except OSError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={"code": "release_workspace_unavailable"},
        ) from exc
    return EditionReleaseMaterializationView(edition_id=edition_id, materialized=True)


@router.get("/editions/{edition_id}/release/docx")
async def download_edition_docx(edition_id: UUID, request: Request) -> Response:
    try:
        _, content = await _publication_service(request).read_docx(edition_id)
    except Exception as exc:
        _raise_publication_error(exc)
    return Response(
        content=content,
        media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        headers={"Content-Disposition": f'attachment; filename="bulletin-{edition_id}.docx"'},
    )


async def _decide(
    edition_id: UUID,
    subject_id: UUID,
    payload: IncludeReviewRequest | ExcludeReviewRequest,
    request: Request,
    decision: PublicationDecision,
) -> ReviewDecisionView:
    try:
        reason = getattr(payload, "reason", None)
        event = await _service(request).decide(
            edition_id,
            subject_id,
            decision=decision,
            production_run_id=payload.production_run_id,
            pipeline_generation=payload.pipeline_generation,
            document_artifact_id=payload.document_artifact_id,
            document_artifact_version=payload.document_artifact_version,
            document_input_hash=payload.document_input_hash,
            actor_id=await _actor_id(request),
            reason=reason,
        )
        return _decision_view(event)
    except Exception as exc:
        _raise_review_error(exc)


def _review_view(review: EditionReview) -> EditionReviewView:
    return EditionReviewView(
        edition_id=review.edition_id,
        items=[
            ReviewItemView(
                position=item.position,
                subject_id=item.subject_id,
                title=item.title,
                run_id=item.run_id,
                pipeline_generation=item.pipeline_generation,
                run_status=item.run_status.value,
                document_artifact_id=item.document_artifact_id,
                document_artifact_version=item.document_artifact_version,
                document_input_hash=item.document_input_hash,
                error_code=item.error_code,
                error_message=item.error_message,
                effective_decision=item.effective_decision,
                effective_decision_id=item.effective_decision_id,
                included=item.included,
                blocking=item.blocking,
                rejected_indicator_count=item.rejected_indicator_count,
                rejected_rule_count=item.rejected_rule_count,
                published_rule_count=item.published_rule_count,
                can_retry=item.can_retry,
                retry_stage=item.retry_stage,
                requires_reconciliation=item.requires_reconciliation,
                reconciliation=reconciliation_view(
                    item.run_id,
                    item.reconciliation,
                    pipeline_generation=item.pipeline_generation,
                ),
            )
            for item in review.items
        ],
        can_accept=review.can_accept,
    )


def _decision_view(decision: PublicationReviewDecision) -> ReviewDecisionView:
    return ReviewDecisionView(
        id=decision.id,
        edition_id=decision.edition_id,
        subject_id=decision.subject_id,
        production_run_id=decision.production_run_id,
        pipeline_generation=decision.pipeline_generation,
        document_artifact_id=decision.document_artifact_id,
        document_artifact_version=decision.document_artifact_version,
        document_input_hash=decision.document_input_hash,
        decision=decision.decision,
        actor_id=decision.actor_id,
        reason=decision.reason,
        occurred_at=decision.occurred_at.isoformat(),
    )


def _release_view(release: EditionReleaseStatus) -> EditionReleaseView:
    return EditionReleaseView(
        edition_id=release.edition_id,
        edition_status=release.edition_status.value,
        manifest_id=release.manifest_id,
        manifest_sha256=release.manifest_sha256,
        release_id=release.release.id if release.release is not None else None,
        json_available=release.json_available,
        markdown_available=release.markdown_available,
        docx_available=release.docx_available,
        published_at=release.published_at.isoformat() if release.published_at else None,
        assembly_job_id=release.assembly_job_id,
        assembly_status=release.assembly_status,
        assembly_error_code=release.assembly_error_code,
        assembly_error_message=release.assembly_error_message,
        can_retry_assembly=release.can_retry_assembly,
    )


def _raise_review_error(exc: Exception) -> NoReturn:
    if isinstance(exc, EditionReviewNotFoundError):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Edition not found"
        ) from exc
    if isinstance(exc, EditionReviewItemNotFoundError):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Review item not found"
        ) from exc
    if isinstance(exc, ReviewItemStaleError):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={"code": "review_item_stale"},
        ) from exc
    if isinstance(exc, EditionReviewStatusError):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={"code": "edition_must_be_in_review"},
        ) from exc
    if isinstance(exc, InvalidReviewReasonError):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={"code": "invalid_review_reason"},
        ) from exc
    if isinstance(exc, InvalidReviewDocumentError):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={"code": "invalid_review_document"},
        ) from exc
    raise exc


def _raise_publication_error(exc: Exception) -> NoReturn:
    if isinstance(exc, PublicationManifestNotFoundError):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Edition release not available"
        ) from exc
    if isinstance(exc, PublicationAcceptanceError):
        code = str(exc)
        if code == "edition_not_found":
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Edition not found"
            ) from exc
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail={"code": code}) from exc
    raise exc

File: frontend/src/api/publication.ts
import { ApiError, type EditionStatus } from "./editions";
import type { ProductionReconciliation } from "./production";

export type { ProductionReconciliation };

export type PublicationDecision = "include" | "exclude";

export type ReviewRunStatus =
  "queued" | "running" | "ready" | "needs_review" | "failed" | "cancelled";

export type ReviewRetryStage =
  "sources" | "references" | "extraction" | "synthesis" | "assembly";

export type ProductionRepairIssueKind =
  "rejected_indicator" | "rejected_rule" | "supplemental_source_unarchived";

export type ProductionRepairAction =
  "include" | "exclude" | "continue_without_source";

/**
 * Durable state of a Q1 proposal missing from the current canonical
 * ReferenceReport. `archived_pending_references` means the analyst already
 * supplied the content and only the deterministic rebuild is missing: the
 * backend keeps that debt, so a refresh can never erase it.
 */
export type SupplementalSourceRepairState =
  "unarchived" | "collection_missing" | "archived_pending_references";

export type AssemblyJobStatus =
  "queued" | "running" | "waiting_human" | "succeeded" | "failed" | "cancelled";

export interface PublicationAcceptResponse {
  edition_id: string;
  edition_status: EditionStatus;
  manifest_id: string;
  manifest_sha256: string;
  edition_version: number;
  batch_id: string;
  job_id: string | null;
  job_dispatched: boolean;
}

export interface EditionReleaseResponse {
  edition_id: string;
  edition_status: EditionStatus;
  manifest_id: string | null;
  manifest_sha256: string | null;
  release_id: string | null;
  json_available: boolean;
  markdown_available: boolean;
  docx_available: boolean;
  published_at: string | null;
  assembly_job_id: string | null;
  assembly_status: AssemblyJobStatus | null;
  assembly_error_code: string | null;
  assembly_error_message: string | null;
  can_retry_assembly: boolean;
}

export interface ReviewItem {
  position: number;
  subject_id: string;
  title: string;
  run_id: string;
  pipeline_generation: number;
  run_status: ReviewRunStatus;
  document_artifact_id: string | null;
  document_artifact_version: number | null;
  document_input_hash: string | null;
  effective_decision_id: string | null;
  effective_decision: PublicationDecision | null;
  included: boolean;
  blocking: boolean;
  rejected_indicator_count: number;
  rejected_ioc_count?: number;
  rejected_other_artifact_count?: number;
  rejected_rule_count: number;
  published_rule_count: number;
  active_repair_count?: number;
  unresolved_repair_count?: number;
  /**
   * The backend owns the Review action policy. `can_retry` and
   * `requires_reconciliation` are mutually exclusive, so the UI never has to
   * infer which action is legal from an error message.
   */
  can_retry: boolean;
  retry_stage: ReviewRetryStage | null;
  requires_reconciliation: boolean;
  reconciliation: ProductionReconciliation | null;
  error_code: string | null;
  error_message: string | null;
}

export type EditionReviewItem = ReviewItem;

export interface EditionReview {
  edition_id: string;
  items: ReviewItem[];
  can_accept: boolean;
  unresolved_repair_count?: number;
  repair_review_complete?: boolean;
  pending_rebuild_count?: number;
}

export interface ProductionRepairDecision {
  id: string;
  action: ProductionRepairAction;
  actor_id: string;
  reason: string | null;
  created_at: string;
  observed_artifact_id: string;
  observed_pipeline_generation: number;
}

export interface EditionRepairItem {
  repair_key: string;
  kind: ProductionRepairIssueKind;
  position: number;
  subject_id: string;
  article_title: string;
  run_id: string;
  pipeline_generation: number;
  artifact_id: string | null;
  artifact_version: number | null;
  source_id: string | null;
  source_title: string | null;
  source_url: string | null;
  collection_id: string | null;
  collection_state: string | null;
  artifact_type: string | null;
  preview: string;
  reason_code: string;
  value_sha256: string;
  payload_available: boolean;
  effective_action: ProductionRepairAction | null;
  effective_decision_id: string | null;
  resolved: boolean;
  resolution_reason: string | null;
  rebuild_required: boolean;
  recommended_stage: string | null;
  repair_state?: SupplementalSourceRepairState | null;
  is_publication_ioc: boolean;
  /** False when the analyst excluded the article from the deliverable. */
  in_publication_scope?: boolean;
}

export interface EditionRepairSummary {
  unresolved_total: number;
  sources_to_supply: number;
  rejected_iocs_to_review: number;
  rejected_rules_to_review: number;
  rejected_other_artifacts: number;
  articles_with_repairs: number;
  articles_needing_rebuild: number;
}

export interface EditionRepairArticle {
  subject_id: string;
  has_pending_projection: boolean;
  recommended_stage: string;
  active_repair_count: number;
  resolved_since_last_build_count: number;
}

export interface EditionRepairPage {
  summary: EditionRepairSummary;
  items: EditionRepairItem[];
  articles: EditionRepairArticle[];
  next_cursor: string | null;
}

export interface EditionRepairDetail {
  repair_key: string;
  kind: ProductionRepairIssueKind;
  artifact_id?: string | null;
  artifact_version?: number | null;
  source_id: string | null;
  source_title: string | null;
  source_url: string | null;
  publisher?: string | null;
  artifact_type?: string | null;
  reason_code?: string | null;
  value_sha256?: string | null;
  preview?: string | null;
  payload_available?: boolean;
  value?: string | null;
  body?: string | null;
  collection_id?: string | null;
  collection_state?: string | null;
  repair_state?: SupplementalSourceRepairState | null;
  rebuild_required?: boolean;
  recommended_action?: string | null;
  effective_decision?: ProductionRepairDecision | null;
  /**
   * Complete append-only audit, oldest first. A decision is revisable, so the
   * last entry is the effective one and the earlier ones explain the change.
   */
  decision_history?: ProductionRepairDecision[];
}

export interface EditionRepairSourcePreparation {
  repair_key: string;
  subject_id: string;
  collection_id: string;
  collection_state: string;
  source_url: string;
}

export interface EditionRepairDecisionResponse {
  repair_key: string;
  decision_id: string;
  action: ProductionRepairAction;
  resolved: true;
}

export interface EditionRepairBulkDecisionResponse {
  decision_ids: string[];
  decisions: Array<{
    repair_key: string;
    decision_id: string;
    action: ProductionRepairAction;
  }>;
}

export interface EditionRepairRebuildResponse {
  action: string;
  stage: string | null;
  run_id: string;
  batch_id: string | null;
  changed: boolean;
  job_id: string | null;
}

export interface ReviewDecision {
  id: string;
  edition_id: string;
  subject_id: string;
  production_run_id: string;
  pipeline_generation: number;
  document_artifact_id: string | null;
  document_artifact_version: number | null;
  document_input_hash: string | null;
  decision: PublicationDecision;
  actor_id: string;
  reason: string | null;
  occurred_at: string;
}

export interface RetryProductionRunResponse {
  run_id: string;
  status: ReviewRunStatus;
  job_id: string | null;
  pipeline_generation: number;
}

export interface CancelProductionRunResponse {
  action: "cancel";
  run_id: string;
  status: "cancelled";
}

interface ReviewDocumentIdentity {
  document_artifact_id: string;
  document_artifact_version: number;
  document_input_hash: string;
}

function documentIdentity(item: ReviewItem): ReviewDocumentIdentity {
  if (
    item.document_artifact_id === null ||
    item.document_artifact_version === null ||
    item.document_input_hash === null
  ) {
    throw new Error("Aucun document publiable n’est disponible.");
  }
  return {
    document_artifact_id: item.document_artifact_id,
    document_artifact_version: item.document_artifact_version,
    document_input_hash: item.document_input_hash,
  };
}

export async function getEditionReview(
  editionId: string,
): Promise<EditionReview> {
  return request(`/api/editions/${editionId}/review`);
}

export async function getEditionRepairPage(
  editionId: string,
  options: {
    status?: "open" | "resolved" | "all";
    kind?: ProductionRepairIssueKind;
    subjectId?: string;
    artifactType?: string;
    cursor?: string | null;
    limit?: number;
  } = {},
): Promise<EditionRepairPage> {
  const params = new URLSearchParams({
    status: options.status ?? "all",
    limit: String(options.limit ?? 100),
  });
  if (options.kind) params.set("kind", options.kind);
  if (options.subjectId) params.set("subject_id", options.subjectId);
  if (options.artifactType) params.set("artifact_type", options.artifactType);
  if (options.cursor) params.set("cursor", options.cursor);
  const payload = await request<unknown>(
    `/api/editions/${encodeURIComponent(editionId)}/review/repairs?${params.toString()}`,
  );
  return isEditionRepairPage(payload) ? payload : emptyEditionRepairPage();
}

export function getEditionRepairDetail(
  editionId: string,
  repairKey: string,
): Promise<EditionRepairDetail> {
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/review/repairs/${encodeURIComponent(repairKey)}`,
  );
}

export function decideEditionRepair(
  editionId: string,
  repairKey: string,
  input: {
    action: ProductionRepairAction;
    observedSubjectId: string;
    observedRunId: string;
    observedArtifactId: string;
    observedPipelineGeneration: number;
    /**
     * Optimistic fence: null for a first decision, the id of the effective
     * decision currently displayed when the analyst revises it.
     */
    expectedEffectiveDecisionId: string | null;
    reason?: string | null;
  },
): Promise<EditionRepairDecisionResponse> {
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/review/repairs/${encodeURIComponent(repairKey)}/decision`,
    jsonRequest({
      action: input.action,
      observed_subject_id: input.observedSubjectId,
      observed_run_id: input.observedRunId,
      observed_artifact_id: input.observedArtifactId,
      observed_pipeline_generation: input.observedPipelineGeneration,
      expected_effective_decision_id: input.expectedEffectiveDecisionId,
      ...(input.reason?.trim() ? { reason: input.reason.trim() } : {}),
    }),
  );
}

/** Create or return the SourceCollection matching one raw Q1 proposal. */
export function prepareEditionRepairSource(
  editionId: string,
  repairKey: string,
): Promise<EditionRepairSourcePreparation> {
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/review/repairs/${encodeURIComponent(repairKey)}/source`,
    { method: "POST" },
  );
}

export function decideEditionRepairsBulk(
  editionId: string,
  decisions: ReadonlyArray<{
    repairKey: string;
    action: ProductionRepairAction;
    observedSubjectId: string;
    observedRunId: string;
    observedArtifactId: string;
    observedPipelineGeneration: number;
    expectedEffectiveDecisionId: string | null;
  }>,
  reason?: string | null,
): Promise<EditionRepairBulkDecisionResponse> {
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/review/repairs/decisions`,
    jsonRequest({
      decisions: decisions.map((decision) => ({
        repair_key: decision.repairKey,
        action: decision.action,
        observed_subject_id: decision.observedSubjectId,
        observed_run_id: decision.observedRunId,
        observed_artifact_id: decision.observedArtifactId,
        observed_pipeline_generation: decision.observedPipelineGeneration,
        expected_effective_decision_id: decision.expectedEffectiveDecisionId,
      })),
      ...(reason?.trim() ? { reason: reason.trim() } : {}),
    }),
  );
}

export function rebuildEditionReviewItem(
  editionId: string,
  subjectId: string,
  observed?: {
    runId?: string;
    pipelineGeneration?: number;
    artifactId?: string;
  },
): Promise<EditionRepairRebuildResponse> {
  const body = observed
    ? {
        ...(observed.runId ? { observed_run_id: observed.runId } : {}),
        ...(observed.pipelineGeneration !== undefined
          ? { observed_pipeline_generation: observed.pipelineGeneration }
          : {}),
        ...(observed.artifactId
          ? { observed_artifact_id: observed.artifactId }
          : {}),
      }
    : undefined;
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/review/items/${encodeURIComponent(subjectId)}/rebuild`,
    body ? jsonRequest(body) : { method: "POST" },
  );
}

export function acceptEditionPublication(
  editionId: string,
): Promise<PublicationAcceptResponse> {
  return request(
    `/api/editions/${encodeURIComponent(editionId)}/publication/accept`,
    {
      method: "POST",
    },
  );
}

export function getEditionRelease(
  editionId: string,
): Promise<EditionReleaseResponse> {
  return request(`/api/editions/${encodeURIComponent(editionId)}/release`);
}

export function editionDocxUrl(editionId: string): string {
  return `/api/editions/${encodeURIComponent(editionId)}/release/docx`;
}

export async function includeReviewItem(
  editionId: string,
  item: ReviewItem,
  reason?: string,
): Promise<ReviewDecision> {
  const identity = documentIdentity(item);
  const normalizedReason = reason?.trim();
  return request(
    `/api/editions/${editionId}/review/items/${item.subject_id}/include`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        production_run_id: item.run_id,
        pipeline_generation: item.pipeline_generation,
        ...identity,
        ...(normalizedReason ? { reason: normalizedReason } : {}),
      }),
    },
  );
}

export async function excludeReviewItem(
  editionId: string,
  item: ReviewItem,
  reason: string,
): Promise<ReviewDecision> {
  return request(
    `/api/editions/${editionId}/review/items/${item.subject_id}/exclude`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        production_run_id: item.run_id,
        pipeline_generation: item.pipeline_generation,
        document_artifact_id: item.document_artifact_id,
        document_artifact_version: item.document_artifact_version,
        document_input_hash: item.document_input_hash,
        reason: reason.trim(),
      }),
    },
  );
}

export async function retryProductionRun(
  runId: string,
  stage: ReviewRetryStage,
): Promise<RetryProductionRunResponse> {
  return request(`/api/production/runs/${runId}/retry`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ stage }),
  });
}

export async function cancelProductionRun(
  runId: string,
): Promise<CancelProductionRunResponse> {
  return request(`/api/production/runs/${runId}/cancel`, {
    method: "POST",
  });
}

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init);
  if (response.ok) return (await response.json()) as T;
  throw await apiError(response);
}

function jsonRequest(body: object): RequestInit {
  return {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  };
}

function isEditionRepairPage(value: unknown): value is EditionRepairPage {
  if (typeof value !== "object" || value === null) return false;
  const record = value as Record<string, unknown>;
  return (
    Array.isArray(record.items) &&
    Array.isArray(record.articles) &&
    typeof record.summary === "object" &&
    record.summary !== null
  );
}

function emptyEditionRepairPage(): EditionRepairPage {
  return {
    summary: {
      unresolved_total: 0,
      sources_to_supply: 0,
      rejected_iocs_to_review: 0,
      rejected_rules_to_review: 0,
      rejected_other_artifacts: 0,
      articles_with_repairs: 0,
      articles_needing_rebuild: 0,
    },
    items: [],
    articles: [],
    next_cursor: null,
  };
}

async function apiError(response: Response): Promise<ApiError> {
  const body = (await response.json().catch(() => null)) as {
    detail?: { code?: unknown; message?: unknown } | string;
  } | null;
  const detail = body?.detail;
  const message =
    typeof detail === "string"
      ? detail
      : typeof detail?.message === "string"
        ? detail.message
        : "La revue de publication n’a pas pu être mise à jour.";
  const code =
    typeof detail === "object" &&
    detail !== null &&
    typeof detail.code === "string"
      ? detail.code
      : "publication_error";
  return new ApiError(message, code, response.status);
}
